flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectResultFetcher.java

```diff
@@ -0,0 +1,359 @@
+			Thread.sleep(retryMillis);
+		} catch (InterruptedException e) {
+			LOG.warn("Interrupted when sleeping before a retry", e);
+		}
+	}
+
+	private void checkJobClientConfigured() {
+		Preconditions.checkNotNull(jobClient, "Job client must be configured before first use.");
+	}
+
+	/**
+	 * A buffer which encapsulates the logic of dealing with the response from the {@link CollectSinkFunction}.
+	 * See Java doc of {@link CollectSinkFunction} for explanation of this communication protocol.
+	 */
+	private class ResultBuffer {
+
+		private static final String INIT_VERSION = "";
+
+		private final LinkedList<T> buffer;
+		private final TypeSerializer<T> serializer;
+
+		// for detailed explanation of the following 3 variables, see Java doc of CollectSinkFunction
+		// `version` is to check if the sink restarts
+		private String version;
+		// `offset` is the offset of the next result we want to fetch
+		private long offset;
+
+		// userVisibleHead <= user visible results offset < userVisibleTail
+		private long userVisibleHead;
+		private long userVisibleTail;
+
+		private ResultBuffer(TypeSerializer<T> serializer) {
+			this.buffer = new LinkedList<>();
+			this.serializer = serializer;
+
+			this.version = INIT_VERSION;
+			this.offset = 0;
+
+			this.userVisibleHead = 0;
+			this.userVisibleTail = 0;
+		}
+
+		private T next() {
+			if (userVisibleHead == userVisibleTail) {
+				return null;
+			}
+			T ret = buffer.removeFirst();
+			userVisibleHead++;
+
+			sanityCheck();
+			return ret;
+		}
+
+		private void dealWithResponse(CollectCoordinationResponse<T> response) {
+			dealWithResponse(response, offset);
+		}
+
+		private void dealWithResponse(CollectCoordinationResponse<T> response, long responseOffset) {
+			String responseVersion = response.getVersion();
+			long responseLastCheckpointedOffset = response.getLastCheckpointedOffset();
+			List<T> results;
+			try {
+				results = response.getResults(serializer);
+			} catch (IOException e) {
+				throw new RuntimeException(e);
+			}
+
+			// we first check version in the response to decide whether we should throw away dirty results
+			if (!version.equals(responseVersion)) {
+				// sink restarted, we revert back to where the sink tells us
+				for (long i = 0; i < offset - responseLastCheckpointedOffset; i++) {
+					buffer.removeLast();
+				}
+				version = responseVersion;
+				offset = responseLastCheckpointedOffset;
+			}
+
+			// we now check if more results can be seen by the user
+			if (responseLastCheckpointedOffset > userVisibleTail) {
+				// lastCheckpointedOffset increases, this means that more results have been
+				// checkpointed, and we can give these results to the user
+				userVisibleTail = responseLastCheckpointedOffset;
+			}
+
+			if (!results.isEmpty()) {
+				// response contains some data, add them to buffer
+				int addStart = (int) (offset - responseOffset);
+				List<T> addedResults = results.subList(addStart, results.size());
```

Comment:
This is impossible. Because if this happens then the sink must have restarted, and we must have gone back to the last checkpointed offset. But I'll add a sanity check here.